### PR TITLE
Remove some allocation from NotifyCollectionChangedEventArgs

### PR DESCRIPTION
**Pull request by stephentoub**
When initialized with a single object rather than a list of objects (single object is very common), the args is allocating an object[] and then wrapping that object[] in a ReadOnlyList.  We can instead just allocate a simple read-only IList wrapper for the object directly; that also makes accessing it faster.

Additionally, when constructing an instance using `public NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction action, IList? changedItems, int index, int oldIndex)`, even though the same collection instance is used for both the new and old items, it wraps it twice.  We can wrap it once.

Along the way I also simplified the code, enabling the fields to become readonly, removing duplicate assignments of fields, etc.

### CODE REVIEW

File: src/libraries/System.ObjectModel/src/System/Collections/CollectionHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+
+namespace System.Collections
+{
+    internal static class CollectionHelpers
+    {
+        internal static void ValidateArguments(int sourceCount, Array array, int index)
```
Comment (eiriktsarpalis) [inline src/libraries/System.ObjectModel/src/System/Collections/CollectionHelpers.cs:10]: Nit: naming feels kind of ambiguous to me. How about `ValidateTargetArray` or `ValidateCopyToArguments`?

Comment (eiriktsarpalis) [reply src/libraries/System.ObjectModel/src/System/Collections/CollectionHelpers.cs:10]: Also, this looks like a good candidate for reuse in System.Collections.Generic implementations.

Comment (stephentoub) [reply src/libraries/System.ObjectModel/src/System/Collections/CollectionHelpers.cs:10]: > Also, this looks like a good candidate for reuse in System.Collections.Generic implementations.

Yes, there are a bunch of places across a variety of projects that could consolidate to use the same function.  I'd prefer we do that separately.

Comment (stephentoub) [reply src/libraries/System.ObjectModel/src/System/Collections/CollectionHelpers.cs:10]: > naming feels kind of ambiguous to me. How about ValidateTargetArray or ValidateCopyToArguments?

Ok, I'll change it to ValidateCopyToArguments.

**Review by eiriktsarpalis — APPROVED** [review]
Thanks

**Pull request merged**